controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -190,222 +188,149 @@ public void cancelChange(ApplicationId applicationId, boolean keepApplicationCha
+                for (JobType job : remainingJobs)
+                    triggerings.add(new Triggering(app, job, reason, stepJobs));
+                completedAt = Optional.empty();
+            }
         }
-    }
+        if (completedAt.isPresent())
+            application = application.withChange(Change.empty());
 
-    private ApplicationController applications() { return controller.applications(); }
+        for (Triggering triggering : triggerings)
+            if (application.deploymentJobs().isDeployableTo(triggering.jobType.environment(), change) && allowedToTriggerNow(triggering, application))
+                application = trigger(triggering, application);
 
-    /** Returns whether the given job type should be triggered according to deployment spec */
-    private boolean hasJob(JobType jobType, Application application) {
-        if ( ! jobType.isProduction()) return true; // Deployment spec only determines this for production jobs.
-        return application.deploymentSpec().includes(jobType.environment(), jobType.region(controller.system()));
-    }
-    /** Create application version from job report */
-    private ApplicationVersion applicationVersionFrom(JobReport report) {
-        return report.sourceRevision().map(sr -> ApplicationVersion.from(sr, report.buildNumber()))
-                     .orElse(ApplicationVersion.unknown);
+        applications().store(application);
     }
 
-    /** Returns true if the given proposed job triggering should be effected */
-    private boolean allowedTriggering(JobType jobType, LockedApplication application) {
-        // Note: We could make a more fine-grained and more correct determination about whether to block
-        //       by instead basing the decision on what is currently deployed in the zone. However,
-        //       this leads to some additional corner cases, and the possibility of blocking an application
-        //       fix to a version upgrade, so not doing it now
-
-        if (jobType.isProduction() && application.change().isPresent() &&
-            application.change().blockedBy(application.deploymentSpec(), clock.instant())) return false;
+    private Optional<Instant> completedAt(Application application, JobType jobType) {
+        return jobType.isProduction()
+                ? changeCompletedAt(application, jobType)
+                : application.deploymentJobs().successAt(application.change(), jobType);
+    }
 
-        // Don't downgrade or redeploy the same version in production needlessly
-        if (jobType.isProduction() && changeDeployed(application, jobType)) return false;
+    private boolean allowedToTriggerNow(Triggering triggering, Application application) {
+        if (application.deploymentJobs().isRunning(triggering.jobType, jobTimeoutLimit()))
+            return false;
 
-        if (application.deploymentJobs().isRunning(jobType, jobTimeoutLimit())) return false;
-        if  ( ! hasJob(jobType, application)) return false;
```

Comment:
This could only stem from a user forced trigger, and must be handled at deploy time, since those jobs can be triggered anyway.